contracts/storage/ServiceAgreementStorage.sol

```diff
@@ -0,0 +1,503 @@
+            CommitSubmission memory commit = commitSubmissions[nextCommitId];
+
+            if (identityId == commit.identityId) isRewarded = true;
+
+            nextCommitId = keccak256(abi.encodePacked(agreementId, epoch, commit.nextIdentity));
+        }
+
+        require(!isRewarded, "You hasn't been chosen for reward in this epoch!");
+
+        bytes32 merkleRoot;
+        uint256 challenge;
+        (merkleRoot, challenge) = getChallenge(assetContract, tokenId, keyword, hashingFunctionId);
+
+        require(
+            MerkleProof.verify(
+                proof,
+                merkleRoot,
+                keccak256(abi.encodePacked(chunkHash, challenge))
+            ),
+            "Root hash doesn't match"
+        );
+
+        ProfileStorage profileStorage = ProfileStorage(hub.getContractAddress("ProfileStorage"));
+
+        // emit ProofSubmitted(
+        //     assetContract,
+        //     tokenId,
+        //     keyword,
+        //     hashingFunctionId,
+        //     identityId,
+        //     profileStorage.getNodeId(identityId)
+        // );
+
+        uint96 reward = serviceAgreements[agreementId].tokenAmount / (serviceAgreements[agreementId].epochsNum - epoch + 1) / (parametersStorage.R0() - serviceAgreements[agreementId].rewardedNodes[epoch]);
+
+        IERC20(hub.getContractAddress("Token")).transfer(address(profileStorage), reward);
+
+        profileStorage.setReward(identityId, profileStorage.getReward(identityId) + reward);
+
+        serviceAgreements[agreementId].tokenAmount -= reward;
+        serviceAgreements[agreementId].rewardedNodes[epoch] += 1;
+
+        // To make sure that node already received reward
+        commitSubmissions[keccak256(abi.encodePacked(agreementId, epoch, identityId))].score = 0;
+    }
+
+    function setScoringFunction(bytes32 agreementId, uint8 newScoringFunctionId)
+        public
+        onlyAssetContracts
+    {
+        serviceAgreements[agreementId].scoringFunctionId = newScoringFunctionId;
+    }
+
+    function _insertCommitAfter(bytes32 agreementId, uint16 epoch, uint96 prevIdentityId, CommitSubmission memory commit)
+        private
+    {
+        bytes32 commitId = keccak256(abi.encodePacked(agreementId, epoch, commit.identityId));
+
+        // Replacing head
+        if (prevIdentityId == 0) {
+            bytes32 epochSubmissionsHead = serviceAgreements[agreementId].epochSubmissionHeads[epoch];
+
+            uint96 prevHeadIdentityId = 0;
+            if(epochSubmissionsHead != "") {
+                CommitSubmission memory commitHead = commitSubmissions[epochSubmissionsHead];
+                prevHeadIdentityId = commitHead.identityId;
+
+                require(
+                    commit.score > commitHead.score,
+                    "Score of the commit must be higher that the score of the head in order to replace it!"
+                );
+            }
+            else if (epoch > 1) {
```

Comment:
lets discuss this live, but I think we should opt for removing / refactoring this else section